### PR TITLE
Offchain-worker APIs stubs

### PR DESCRIPTION
**Pull request by tomusdrw**
This PR introduces a new APIs avaiable for offchain workers, although they are not implemented yet (this will follow in subsequent PRs and can be parallelized)

A list of APIs include:
- `submit_transaction`
- `random_seed`
- `sign`
- `local_storage_read/write`
- HTTP apis

This PR is to discuss the proposed APIs and the implementation of wasm-native boundaries.

Related: #1458

### CODE REVIEW

File: core/sr-io/without_std.rs [hunk 1]
```diff
@@ -624,10 +718,200 @@ impl CryptoApi for () {
...
+	}
+
+	fn http_request_start(method: &str, url: &str, meta: &[u8]) -> Result<offchain::HttpRequestId, ()> {
+		let method = method.as_bytes();
+		let url = url.as_bytes();
+
+		unsafe {
+			let result = ext_http_request_start.get()(
+				method.as_ptr(),
+				method.len() as u8,
```
Comment (pepyakin) [inline core/sr-io/without_std.rs]: Although truncation here cannot lead to unsafety, maybe it would still be good to return `Err` if the conversion to `u8` fails?

File: core/sr-io/without_std.rs [hunk 2]
```diff
@@ -331,8 +331,102 @@ pub mod ext {
...
+		///
+		/// if len is u32::max_value the value has not been found.
+		fn ext_local_storage_read(key: *const u8, key_len: u32, value_len: *mut u32) -> *mut u8;
+
+		/// Initiaties a http request.
+		///
+		/// If returned value is `u32::max_value` the request couldn't have started.
+		fn ext_http_request_start(
+			method: *const u8,
+			method_len: u8,
```
Comment (pepyakin) [inline core/sr-io/without_std.rs]: I am not entirely sure why would we want to use `u8` here?

On wasm32 platforms it would be passed as `i32` anyway. I am not entirely convienced that the optimizer can take advantage of `u8` and that it would actually be benefital.

Consider changing it to u32.

File: core/sr-io/without_std.rs [hunk 3]
```diff
@@ -331,8 +331,102 @@ pub mod ext {
...
+		fn ext_local_storage_set(key: *const u8, key_len: u32, value: *const u8, value_len: u32);
+
+		/// Read a value from local storage.
+		///
+		/// if len is u32::max_value the value has not been found.
+		fn ext_local_storage_read(key: *const u8, key_len: u32, value_len: *mut u32) -> *mut u8;
+
+		/// Initiaties a http request.
+		///
+		/// If returned value is `u32::max_value` the request couldn't have started.
```
[the remark this answers is not in the document]
Comment (pepyakin) [reply core/sr-io/without_std.rs]: I am not sure about `u32::max_value` comment. It actually looks for `u16`.

File: core/sr-io/without_std.rs [hunk 4]
```diff
@@ -624,10 +718,200 @@ impl CryptoApi for () {
...
+		unsafe {
+			ext_http_response_wait.get()(
+				ids.as_ptr(),
+				ids.len() as u32,
+				statuses.as_mut_ptr(),
+				deadline.map_or(0, |x| x.unix_millis()),
+			)
+		}
+
+		statuses
```
Comment (pepyakin) [inline core/sr-io/without_std.rs]: This line exceeds the 120 limit. Can we wrap this line?

File: core/executor/src/wasm_executor.rs [hunk 5]
```diff
@@ -589,15 +599,236 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
...
+					.map_err(|_| UserError("OOB while ext_http_response_wait: wasm"))
+			)
+			.collect::<::std::result::Result<Vec<_>, _>>()?;
+
+		let res = this.ext.offchain()
+			.map(|api| api.http_response_wait(&ids, deadline_to_timestamp(deadline)))
+			.ok_or_else(|| UserError("Calling unavailable API ext_http_request_write_body: wasm"))?
+			.into_iter()
+			.map(|status| status.as_u32())
+			.zip(0..ids_len);
```
Comment (pepyakin) [inline core/executor/src/wasm_executor.rs]: I am curious why not enumerate?

File: core/sr-io/without_std.rs [hunk 6]
```diff
@@ -331,8 +331,102 @@ pub mod ext {
...
+		fn ext_http_response_wait(
+			ids: *const u32,
+			ids_len: u32,
+			statuses: *mut u32,
+			deadline: u64
+		);
+
+		/// Read all response headers.
+		///
+		/// Resturns parity-codec encoded vector of pairs `(HeaderKey, HeaderValue)`.
```
Comment (pepyakin) [inline core/sr-io/without_std.rs]:
```suggestion
		/// Returns parity-codec encoded vector of pairs `(HeaderKey, HeaderValue)`.
```

Comment (pepyakin) [reply core/sr-io/without_std.rs]: I am a bit concerned that it doesn't return a `Result` and at the same time doesn't mention any preconditions like we should call this method only if the status of `id` is appropriate.

File: core/sr-io/src/offchain/http.rs [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+	}
+}
+
+#[cfg(feature = "std")]
+fn from_utf8(chunk: &[u8]) -> Option<&str> {
+	std::str::from_utf8(chunk).ok()
+}
+#[cfg(not(feature = "std"))]
+fn from_utf8(chunk: &[u8]) -> Option<&str> {
+	core::str::from_utf8(chunk).ok()
```
Comment (pepyakin) [inline core/sr-io/src/offchain/http.rs]: Oh wow this is spicy. Asks for adding `str` to rstd.

File: core/sr-io/src/offchain/http.rs [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+
+		// start an http request.
+		let id = crate::http_request_start(self.method.as_ref(), self.url, meta)?;
+
+		// add custom headers
+		for (header_name, header_value) in &self.headers {
+			crate::http_request_add_header(
+				id,
+				from_utf8(header_name).expect("Header names are always Vecs created from valid str; qed"),
+				from_utf8(header_value).expect("Header values are always Vecs created from valid str; qed"),
```
Comment (pepyakin) [inline core/sr-io/src/offchain/http.rs]: Instead of converting strings to vectors and back to strings (which is by the way is not 0-cost, since it involves validation that takes linear time) can we store `String` in `self.headers`?

File: core/sr-io/src/offchain/http.rs [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+	pub fn wait(self) -> Result<Response, Error> {
+		match self.try_wait(None) {
+			Ok(res) => res,
+			Err(_) => panic!("Since `None` is passed we will never get a deadline error; qed"),
+		}
+	}
+
+	/// Attempts to wait for the request to finish,
+	/// but will return `Err` in case the deadline is reached.
+	pub fn try_wait(self, deadline: impl Into<Option<Timestamp>>) -> Result<Result<Response, Error>, PendingRequest> {
```
Comment (pepyakin) [inline core/sr-io/src/offchain/http.rs]: Nit: you probably know, but nested results doesn't look that awesome. It is hard to wrap my head around which `Err` serves which purpose...

File: core/primitives/src/offchain.rs [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+	/// If the value does not exist in the storage `None` will be returned.
+	/// Note this storage is not part of the consensus, it's only accessible by
+	/// offchain worker tasks running on the same machine. It IS persisted between runs.
+	fn local_storage_read(&mut self, key: &[u8]) -> Option<Vec<u8>>;
+
+	/// Initiaties a http request given HTTP verb and the URL.
+	///
+	/// Meta is a future-reserved field containing additional, parity-codec encoded parameters.
+	/// Returns the id of newly started request.
+	fn http_request_start(
```
Comment (pepyakin) [inline core/primitives/src/offchain.rs]: I wonder what is the lifetime of a request / response. Looks like I can read headers after the request finished. Does they go away after the off-chain worker finishes (they finish at some point, don't they?)?

File: core/sr-io/src/offchain/http.rs [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+		self.collection.get(self.index.unwrap_or(0))
+			.map(|val| (from_utf8(&val.0).unwrap_or(""), from_utf8(&val.1).unwrap_or("")))
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	#[test]
+	fn write_some() {
+		assert_eq!(true, false)
```
Comment (pepyakin) [inline core/sr-io/src/offchain/http.rs]: Hmm there should be a real test right? : )

File: core/sr-io/src/offchain/http.rs [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+			return None;
+		}
+
+		if self.filled_up_to.is_none() {
+			let result = crate::http_response_read_body(self.id, &mut self.buffer, self.deadline).ok();
+			if let Some(size) = result {
+				self.position = 0;
+				self.filled_up_to = Some(size);
+			} else {
+				self.is_deadline_reached = true;
```
Comment (pepyakin) [inline core/sr-io/src/offchain/http.rs]: Does this only shows reaching the deadline? What about other I/O errors?

File: core/sr-io/src/offchain/http.rs [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Headers {
+	/// Raw headers
+	pub raw: Vec<(Vec<u8>, Vec<u8>)>,
+}
+
+impl Headers {
+	/// Retrieve a single header from the list of headers.
+	///
+	/// Note this method is linearly looking from all the headers.
```
Comment (pepyakin) [inline core/sr-io/src/offchain/http.rs]: doesn't `from_utf8` make it quadratic?

File: core/sr-io/src/offchain/http.rs [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+	/// Returns `true` is `current` has been set by this call.
+	pub fn next(&mut self) -> bool {
+		let index = self.index.map(|x| x + 1).unwrap_or(0);
+		self.index = Some(index);
+		index < self.collection.len()
+	}
+
+	/// Returns current element (if any).
+	///
+	/// Note that you have to call `next` prior to calling this
```
Comment (pepyakin) [inline core/sr-io/src/offchain/http.rs]: Maybe if `next` wasn't called instead of `self.index.unwrap_or(0)` it would be better to return `None`?

I wonder if `self.index?` is possible : )

**Comment by tomusdrw** [comment]
Marking as `inprogress` as I there is quite a lot stuff to fix.

**Comment by gterzian** [comment]
The API reads like it is sync in nature, is that the intention?

**Comment by tomusdrw** [comment]
@gterzian Well, the HTTP API may be used in a non-blocking fashion (but without polling though), the rest is not that important. If we want to maintain WASM compatibility I don't think it's feasible to make it future-based. If we don't need WASM then we probably don't need any of those abstractions actually.

**Comment by gterzian** [comment]
> If we want to maintain WASM compatibility I don't think it's feasible to make it future-based. 

Do you need to maintain wasm compat in the implementation of the API offered to the wasm code, as opposed to just in the API?

File: core/primitives/src/offchain.rs [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+	///
+	/// This is a trully random non deterministic seed generated by host environment.
+	/// Obviously fine in the off-chain worker context.
+	fn random_seed(&mut self) -> [u8; 32];
+
+	/// Sets a value in the local storage.
+	///
+	/// Note this storage is not part of the consensus, it's only accessible by
+	/// offchain worker tasks running on the same machine. It IS persisted between runs.
+	fn local_storage_set(&mut self, key: &[u8], value: &[u8]);
```
Comment (tomaka) [inline core/primitives/src/offchain.rs]: Why is this needed? It feels like introducing a race condition on purpose.

File: core/sr-io/without_std.rs [hunk 16]
```diff
@@ -624,10 +718,200 @@ impl CryptoApi for () {
...
+			let ret = ext_submit_transaction.get()(encoded_data.as_ptr(), encoded_data.len() as u32);
+			if ret == 0 {
+				Ok(())
+			} else {
+				Err(())
+			}
+		}
+	}
+
+	fn sign(data: &[u8]) -> Option<[u8; 64]> {
```
[the remark this answers is not in the document]
Comment (tomusdrw) [reply core/sr-io/without_std.rs]: That's the size of current crypto used (both ed and sr). Do you think it's better (more future proof) to have `Vec<u8>` instead?

File: core/sr-io/without_std.rs (hunk 1 above)
Comment (tomusdrw) [reply core/sr-io/without_std.rs]: Changed to `u32`

File: core/sr-io/src/offchain/http.rs (hunk 8 above)
Comment (pepyakin) [reply core/sr-io/src/offchain/http.rs]: Oh, that's bummer. This clearly demonstrates mismatch between priorities of on-chain and off-chain code. I believe sooner or later off-chain would want to use `String`s anyway. 

I saw your proposed solution. The best thing is to avoid unsafe if we can. If we can't then I think the following solution will be a bit better:

1. introduce a struct, let's call it `Header`.
2. In order to construct an instance of `Header` you have to provide `name: &str` and `value: &str`.
3. the body of the constructor would do `to_bytes().to_vec()`
4. then we add another function that adds this header to the request.

IMO this is better since the code that make guarantees and the code that relies on those guarantees is physically closer to each other. Furthermore, the whole invariant is the almost the single purpose of this struct. These facts should make the proof better, more maintanable and should decrease chances of violating the invariant accidentally with all of this being zero-cost.

wdyt?

File: core/sr-io/src/offchain/http.rs (hunk 13 above)
Comment (pepyakin) [reply core/sr-io/src/offchain/http.rs]: Ah you're right, mb sth like this

```suggestion
	/// Note this method is linearly looking from all the headers comparing them with the needle byte-by-byte.
```

File: core/primitives/src/offchain.rs (hunk 10 above)
Comment (pepyakin) [reply core/primitives/src/offchain.rs]: to clarify: can off-chain workers run indefinitely, like in a while loop?

File: core/primitives/src/offchain.rs [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+pub struct HttpRequestId(pub u16);
+
+/// Status of the HTTP request
+#[derive(Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub enum HttpRequestStatus {
+	/// Deadline was reached why we waited for this request to finish.
+	DeadlineReached,
+	/// Request timed out.
+	Timeout,
```
[the remark this answers is not in the document]
Comment (pepyakin) [reply core/primitives/src/offchain.rs]: Who defines "a reasonable time" (plus typo)

File: core/executor/src/wasm_executor.rs (hunk 5 above)
Comment (pepyakin) [reply core/executor/src/wasm_executor.rs]: I consider it as more readable, yes. Comment would also do.

File: core/sr-io/without_std.rs [hunk 18]
```diff
@@ -331,8 +331,102 @@ pub mod ext {
...
+			deadline: u64
+		) -> u32;
+
+		/// Block and wait for the responses for given requests.
+		///
+		/// Note that if deadline is not provided the method will block indefinitely,
+		/// otherwise unready responses will produce `WaitTimeout` status.
+		///
+		/// Make sure that `statuses` have the same length as ids.
+		/// Passing `0` as deadline blocks forever.
```
[the remark this answers is not in the document]
Comment (pepyakin) [reply core/sr-io/without_std.rs]: To clarify:
- `Note that if deadline is not provided the method will block indefinitely`. 
- `Passing `0` as deadline blocks forever.`

I'd just leave one of them. Up to you which one.

File: core/sr-io/without_std.rs [hunk 19]
```diff
@@ -331,8 +331,116 @@ pub mod ext {
...
+		///
+		/// `data` has to be a pointer to a slice of 32 bytes.
+		fn ext_random_seed(data: *mut u8);
+
+		/// Write a value to local storage.
+		fn ext_local_storage_set(key: *const u8, key_len: u32, value: *const u8, value_len: u32);
+
+		/// Read a value from local storage.
+		///
+		/// if `value_len` is u32::max_value the value has not been found.
```
Comment (pepyakin) [inline core/sr-io/without_std.rs]: I'd like to see what happens with the result value in this case. If the return value is `NULL` in this case then can the user of this code rely this fact instead of checking for `max_value`? If not, why?   This is still unallocated memory then (or is the user expected to call `free` on this explicitly)?

File: core/sr-io/without_std.rs [hunk 20]
```diff
@@ -331,8 +331,116 @@ pub mod ext {
...
+			ids_len: u32,
+			statuses: *mut u32,
+			deadline: u64
+		);
+
+		/// Read all response headers.
+		///
+		/// Note the headers are only available before response body is fully consumed.
+		/// Returns parity-codec encoded vector of pairs `(HeaderKey, HeaderValue)`.
+		/// In case invalid `id` is passed it returns an empty vector.
```
Comment (pepyakin) [inline core/sr-io/without_std.rs]: Tbh, I'd avoid the vector terminology here, because one could mistakingly assume that it is a `Vec` or can be safely treated as so (it can't: see my other response about UB invoked here).

File: core/executor/src/wasm_executor.rs [hunk 21]
```diff
@@ -589,15 +599,238 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
...
+	) -> *mut u8 => {
+		use parity_codec::Encode;
+
+		let headers = this.ext.offchain()
+			.map(|api| api.http_response_headers(offchain::HttpRequestId(request_id as u16)))
+			.ok_or_else(|| UserError("Calling unavailable API ext_http_response_headers: wasm"))?;
+
+		let encoded = headers.encode();
+		let len = encoded.len() as u32;
+		let offset = this.heap.allocate(len)? as u32;
```
Comment (pepyakin) [inline core/executor/src/wasm_executor.rs]: `len` cannot be zero here, right?

File: core/sr-io/without_std.rs [hunk 22]
```diff
@@ -624,10 +732,200 @@ impl CryptoApi for () {
...
+		let mut len = 0u32;
+		let raw_result = unsafe {
+			let ptr = ext_http_response_headers.get()(
+				request_id.0 as u32,
+				&mut len,
+			);
+			// Invariants required by Vec::from_raw_parts are not formally fulfilled.
+			// We don't allocate via String/Vec<T>, but use a custom allocator instead.
+			// See #300 for more details.
+			<Vec<u8>>::from_raw_parts(ptr, len as usize, len as usize)
```
Comment (pepyakin) [inline core/sr-io/without_std.rs]: This is really subtle. `ext_http_response_headers` potentially can invoke UB here, if `ptr` is `null` here, even if `len` equals 0.

We might want to either:

- explicitly mention this here and for the documentation in `ext_http_response_header` and guarantee it
- guard here against this explicitly here.

File: core/sr-io/without_std.rs [hunk 23]
```diff
@@ -624,10 +732,200 @@ impl CryptoApi for () {
...
+			);
+
+			if len == u32::max_value() {
+				None
+			} else {
+
+				// Invariants required by Vec::from_raw_parts are not formally fulfilled.
+				// We don't allocate via String/Vec<T>, but use a custom allocator instead.
+				// See #300 for more details.
+				Some(<Vec<u8>>::from_raw_parts(ptr, len as usize, len as usize))
```
Comment (pepyakin) [inline core/sr-io/without_std.rs]: The same as in https://github.com/paritytech/substrate/pull/2615/files#r287413971

Comment (pepyakin) [reply core/sr-io/without_std.rs]: For efficiency reasons I wonder if we actually want to allocate zero-sized buffers.

File: core/sr-io/src/offchain/http.rs [hunk 24]
```diff
@@ -0,0 +1,400 @@
...
+		// start an http request.
+		let id = crate::http_request_start(self.method.as_ref(), self.url, meta)?;
+
+		// add custom headers
+		for (header_name, header_value) in &self.headers {
+			crate::http_request_add_header(
+				id,
+				// Header (key, values) are always produced from `&str` so this is safe.
+				// we don't store them as `Strings` to avoid bringing `alloc::String` to rstd
+				// or here.
```
Comment (pepyakin) [inline core/sr-io/src/offchain/http.rs]: 👍 thanks for mentioning this!

**Comment by gavofyork** [comment]
Other stuff would include:
- Persistent secret creation and crypto operations for a few cryptos (signing, encryption and decryption). Notably, wasm would never be able to get hold of the secret itself.
- Secure random number generation

File: core/sr-io/without_std.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (tomusdrw) [reply core/sr-io/without_std.rs]: Oh, missed that. Thanks!

File: core/sr-io/without_std.rs (hunk 23 above)
Comment (pepyakin) [reply core/sr-io/without_std.rs]: Yeah, I am ok with that!

File: core/sr-io/without_std.rs (hunk 19 above)
Comment (pepyakin) [reply core/sr-io/without_std.rs]: > So since the docs don't mention anything you can't rely on returned value to be NULL. I can add a clarification comment to all methods that have the same assumptions if you feel it's needed.

That would be great!

File: core/sr-io/without_std.rs (hunk 20 above)
Comment (pepyakin) [reply core/sr-io/without_std.rs]: Yes, thanks!

**Comment by tomusdrw** [comment]
@gavofyork added signing API stubs as well, could you have a look?
Regarding random number generator we already have `ext_random_seed() -> [u8; 32]` It can be either used to seed a generator in WASM or just as source of random numbers. 
I can write some higher level util (like `next_u64()`) that utilizes this API

File: core/sr-io/without_std.rs (hunk 23 above)
Comment (tomusdrw) [reply core/sr-io/without_std.rs]: Logged here: https://github.com/paritytech/substrate/issues/2730

**Comment by gavofyork** [comment]
I don't quite see how `sign` works with taking a `CryptoKeyId` (`pub struct CryptoKeyId(pub u16)`)...

**Comment by tomusdrw** [comment]
`sign(None, data)` returns a signature for data using the current authority (session) key if available.

```
let key = new_crypto_key(CryptoKind::Ed25519);
let signature = sign(Some(key), data);
```
Will return a signature for `data` but using the key created earlier with `new_crypto_key`.
So the `CryptoKeyId` is just some opaque identifier of previously created key.

I've unified the API you've mentioned in previous comment:
> Persistent secret creation and crypto operations for a few cryptos (signing, encryption and decryption). Notably, wasm would never be able to get hold of the secret itself.

and the API to sign a piece of data using current authority (session) key.

Is this what you had in mind?

**Comment by gavofyork** [comment]
I see - fair enough. I had imagined to use the public key to identify keys rather than an opaque ID, but I guess it's fine.

**Comment by gavofyork** [comment]
@pepyakin @tomaka any final remarks?
